docs/ALU Design.md

```diff
@@ -0,0 +1,758 @@
+
+#### Keywords
+
+An instruction is not just its opcode and operands, e.g. `$3 = add $1, $2`, but there are some
+keywords modifying its behavior. An
+[example for `add`](https://llvm.org/docs/LangRef.html#add-instruction):
+
+```llvm
+<result> = add <ty> <op1>, <op2>          ; yields ty:result
+<result> = add nuw <ty> <op1>, <op2>      ; yields ty:result
+<result> = add nsw <ty> <op1>, <op2>      ; yields ty:result
+<result> = add nuw nsw <ty> <op1>, <op2>  ; yields ty:result
+```
+
+- `<ty>` is type, e.g. `i32`,
+- `nuw` - No Unsigned Wrap,
+- `nsw` - No Signed Wrap.
+
+#### Poison
+
+In the example of `add`, if `nuw` or `nsw` keywords occur, they guarantee specific behavior, i.e. no
+(un)signed overflow. However, if the operands cause the overflow, the instruction returns a poison,
+which is an equivalent of a value indicating undefined behavior that can propagate throughout the
+program.
+
+[According to the experiment](https://github.com/reilabs/llvm-to-cairo/issues/27#issuecomment-2397645979),
+LLVM does not seem to emit such instructions from the Rust code, so the initial version of ALU will
+not handle `nuw`, `nsw` or other keywords in any specific way.
+
+### Intrinsics
+
+The example above includes the following line:
+
+```llvm
+%0 = call { i64, i1 } @llvm.uadd.with.overflow.i64(i64 %left, i64 %right), !dbg !17
+```
+
+Unlike the previous example, this snippet does nto contain the `add` instruction. The adding
+operation is done by an intrinsic named `llvm.uadd.with.overflow.i64`, which is called with the
+`call` instruction. The intrinsic is defined in the LLVM codebase and its source code does not make
+it into the `.ll` file produced out of the adding operation in Rust code.
+
+The LLVM Language Reference Manual has an extensive list of intrinsics. Here's the example of
+[`llvm.uadd.with.overflow.<ty>`](https://llvm.org/docs/LangRef.html#llvm-uadd-with-overflow-intrinsics).
+
+### Data types
+
+This ALU will support only integers. Floating point numbers are out of scope of this document.
+
+#### Integers
+
+LLVM IR supports integers of arbitrary width. A general syntax for an integer type is `iN`, where
+`N` can be anything from 1 to 2^32. Since LLVM does not have a dedicated type for boolean values,
+`i1` is used instead.
+
+The Cairo VM internally operates on 252-bit-long field elements - `felt252`. On the higher level of
+abstraction, the Cairo language supports
+[integers of specific lengths](https://book.cairo-lang.org/ch02-02-data-types.html): 8 bit, 16 bit,
+32 bit, 64 bit, 128 bit and 256 bit. Cairo also supports booleans.
+
+[Rust supports integers of width from 8 to 128 bit](https://doc.rust-lang.org/book/ch03-02-data-types.html)
+with the same increment Cairo does, plus architecture-dependent `isize` and `usize`. Rust also
+supports booleans.
+
+The Cairo VM does not have a classical registers of length constrained by the hardware. Therefore
```

Comment:
```suggestion
The Cairo VM does not have classical registers of a length constrained by the hardware. Therefore
```

Technically it does have length-constrained registers, but they are just 252-bit cells instead.